social_core/backends/azuread_b2c.py

```diff
@@ -0,0 +1,181 @@
+    OPENID_CONFIGURATION_URL = '{base_url}/v2.0/.well-known/openid-configuration?p={policy}'
+    ACCESS_TOKEN_URL = '{base_url}/oauth2/v2.0/token?p={policy}'
+    JWKS_URL = '{base_url}/discovery/v2.0/keys?p={policy}'
+
+    DEFAULT_SCOPE = ['openid', 'email']
+
+    EXTRA_DATA = [
+        ('access_token', 'access_token'),
+        ('id_token', 'id_token'),
+        ('refresh_token', 'refresh_token'),
+        ('id_token_expires_in', 'expires'),
+        ('exp', 'expires_on'),
+        ('not_before', 'not_before'),
+        ('given_name', 'first_name'),
+        ('family_name', 'last_name'),
+        ('tfp', 'policy'),
+        ('token_type', 'token_type')
+    ]
+
+    @property
+    def tenant_id(self):
+        return self.setting('TENANT_ID', 'common')
+
+    @property
+    def policy(self):
+        policy = self.setting('POLICY')
+
+        if not policy or not policy.startswith('b2c_'):
+            raise Exception('SOCIAL_AUTH_AZUREAD_B2C_OAUTH2_POLICY is required and should start with `b2c_`')
+
+        return policy
+
+    @property
+    def base_url(self):
+        return self.BASE_URL.format(tenant_id=self.tenant_id)
+
+    def openid_configuration_url(self):
+        return self.OPENID_CONFIGURATION_URL.format(base_url=self.base_url, policy=self.policy)
+
+    def authorization_url(self):
+        # Policy is required, but added later by `auth_extra_arguments()`
+        return self.AUTHORIZATION_URL.format(base_url=self.base_url)
+
+    def access_token_url(self):
+        return self.ACCESS_TOKEN_URL.format(base_url=self.base_url, policy=self.policy)
+
+    def jwks_url(self):
+        return self.JWKS_URL.format(base_url=self.base_url, policy=self.policy)
+
+    def request_access_token(self, *args, **kwargs):
+        """
+        This is probably a hack, but otherwise AzureADOAuth2 expects `access_token`.
+
+        However, B2C backends provides `id_token`.
+        """
+        response = super(AzureADB2COAuth2, self).request_access_token(*args, **kwargs)
+        if 'access_token' not in response:
+            response['access_token'] = response['id_token']
+        return response
+
+    def auth_extra_arguments(self):
+        """
+        Return extra arguments needed on auth process.
+
+        The defaults can be overridden by GET parameters.
+        """
+        extra_arguments = super(AzureADB2COAuth2, self).auth_extra_arguments()
+        extra_arguments['p'] = self.policy or self.data.get('p')
+        return extra_arguments
+
+    def jwt_key_to_pem(self, key_json_dict):
+        """
+        Builds a PEM formatted key string from a JWT public key dict.
+        """
+        pub_key = RSAAlgorithm.from_jwk(json.dumps(key_json_dict))
+        return pub_key.public_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PublicFormat.SubjectPublicKeyInfo
+        )
+
+    def get_user_id(self, details, response):
+        """Use email as unique id"""
```

Comment:
No user id? The `id` is used as the `uid` which is later used to determine if the user already used this provider and authenticate it in that case.

Reply:
I must have forgot that :see_no_evil: Looking at the  This is a sample user details below, and [azure token docs], I believe `sub` is the correct field.

```#python
{
        'aud': 'abcdef12-1234-9876-0000-abcdef098765',
        'auth_time': 1355152329,
        'country': 'Axphain',
        'emails': [
            'foobar@example.com'
        ],
        'exp': 1355155929,
        'family_name': 'Bar',
        'given_name': 'Foo',
        'iat': 1355152329,
        'iss': 'https://login.microsoftonline.com/9a9a9a9a-1111-5555-0000-bc24adfdae00/v2.0/',
        'name': 'FooBar',
        'nbf': 1355152329,
        'oid': '11223344-5566-7788-9999-aabbccddeeff',
        'postalCode': '00000',
        'sub': '11223344-5566-7788-9999-aabbccddeeff',
        'tfp': 'B2C_1_SignIn',
        'ver': '1.0',
}
```